File: 0003.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        pres = dict()
        longest, lpointer = 0, 0
        for i in range(len(s)):
            if s[i] not in pres:
                longest = max(longest, i-lpointer+1)
            else:
                if pres[s[i]] < lpointer:
                    longest = max(longest, i-lpointer+1)
                else:
                    lpointer = pres[s[i]] + 1
            pres[s[i]] = i
        return longest

File: src/0003.cpp
#include <iostream>
#include <unordered_map>

using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> pres;
        int longest = 0, ptr = 0;
        for (int i = 0; i < s.size(); ++i) {
            if (pres.count(s[i]) == 0) {
                longest = max(longest, i-ptr+1);
            }
            else {
                if (pres[s[i]] < ptr) {
                    longest = max(longest, i-ptr+1);
                }
                else {
                    ptr = pres[s[i]] + 1;    
                }
            }
            pres[s[i]] = i;
        }
        return longest;
    }
};
